### DISCUSSION TITLE
Execute SQL in Supabase API

### DISCUSSION THREAD

**Discussion by ltanady** [supabase/supabase#3419] [Questions] [answered]
Is it possible execute SQL in supabase API? I have created a query to group many columns in my table, but I cannot do it using your API, I didn't find any example to extract years or months from dates and neither an example to group columns using SUM or COUNT for example.

**Comment by ltanady** [comment, marked as answer]
We do not allow direct execution of SQL from our APIs. There are two possible solutions:

1. You can create a function with the query specified and make RPC calls as such or
2. You can also generate a view with the query specified, then make a SELECT from there

The latter would be the simpler solution. Hope this helps!

**Reply by steve-chavez** [reply]
Not being able to execute raw SQL is a deliberate decision, SQL cannot be fully exposed to frontend clients because it opens a door for forcing inefficient/slow queries that can starve your database of resources. Also a malicious user could do a Denial-of-Service.

This is why custom SQL should be used behind views or functions - so the developer can ensure the queries are fast/efficient(WHERE conditions are enforced, indexes are applied, etc).

Previous discussion https://github.com/supabase/supabase/discussions/444#discussioncomment-248382

**Reply by ox-harris** [reply]
You know, it would save many of us the wasted hopes if you guys said it upfront in the docs that supabase does not backend/support nodejs-based development. I figured that out in a series of blockers - from not being able to run raw SQL aptly from the backend to not being able to upload a file using the supabase storage API. On running raw SQL, how can you rule out CORRECT USE of the most fundamental way we communicate with SQL? If this whole supabase thing is all about the client libraries (that must also run client-side), it is not for me.

I quit here.

**Reply by steve-chavez** [reply]
> if you guys said it upfront in the docs that supabase does not backend/support nodejs-based development
> from not being able to run raw SQL aptly from the backend

@ox-harris For backend development, you have a raw postgres connection and pgbouncer pooling from where you can connect any postgres library(and do raw SQL) in any language -  you're not forced to use Supabase client libraries.

There are many developers using Supabase with Prisma for example(a quick google search will show you many results).

> I figured that out in a series of blockers

Sorry for the confusion caused. Did you try asking in our [discord](https://discord.supabase.com) first? There's an active community there which can guide you and help you clear any doubts.

**Reply by ox-harris** [reply]
Well, this is what happens with poor documentation. Devs spend their life to dig for information. Why does that important stuff belong in an outside community instead of the official docs?

**Reply by steve-chavez** [reply]
> Why does that important stuff belong in an outside community instead of the official docs?

There are some docs in the works - Prisma(https://github.com/supabase/supabase/pull/3022) and Redwood(https://github.com/supabase/supabase/pull/3573).

**Reply by ox-harris** [reply]
I see. Thanks

**Reply by ForbiddenEra** [reply]
Admittedly, it's even _more_ misleading when you set it up and see that you _can_ do raw SQL in the Supabase UI. I spent at least 5-10 min clicking through different docs pages not finding anything about it and then Google brought me here to be disappointed.

I understand the "reasons" but I didn't particularly get the 'frontend'/'browser-first' connotations when trying this out, even as someone with some previous Firebase experience. I totally agree no one should be doing SQL from a frontend/ui (even on an internal app) but I also don't feel like 100% of Supabase stuff should only be done on the frontend either and never a backend or service.

I found Supabase after seeing a YT video about how you implemented your clippy thing; looked like a pretty simple/quick thing to throw together myself on the weekend and knew it'd at a minimum get a chuckle out of my coworkers.

Following that and [this](https://supabase.com/blog/openai-embeddings-postgres-vector), I got it put together pretty quick. I quickly found out the limitation when I wasn't able to easily run the SQL initialization steps from within code and at first hadn't seen the raw SQL option in the UI leaving me to download a pg client and do it that way which was a bit frustrating as I'd already templated the SQL code w/configuration options.

This is also when I realized; what's the point of even using Supabase here? I get showing off your own product and dogfooding it and considering that the blog was based off your internal implementation but for the purposes of the implementation in the tutorial, from what I can tell, it's using ZERO Supabase-specific features and adding ZERO value in this case, in fact it's almost the opposite, it's only more modules to be loaded in node, more services to be run, etc.

Again I understand dogfooding but even in your own implementation I feel like it would've been simpler to just use a standard pg client library. If there was _any_ Supabase feature or functionality shown off here, I could understand but it seems like even your own implementation is using it unnecessarily here; I get that Supabase is likely already running/setup/etc for other parts but even still, unless I missed something, you could switch over to a standard lib and have zero changes to functionality/features/anything but less resource usage/bloat/modules required for this implementation. 

While I appreciate the blog and the accompanying video as a pretty straight forward explanation and method for implementing this, other than perhaps slightly excessive dogfooding, I don't see the purpose of using it in this specific use-case at all, but I digress: it's your choice how you want to implement things in your own software.

I kept the code close to the provided examples to get my own version up and running as quickly as possible but will now be swapping Supabase out with a standard client library which will also enable the ability of my app/service to initialize the vector database, etc. To be honest, I'd have not bothered with the swap if I could've executed arbitrary SQL, even if that comes with some kind of warning or even limitation preventing it's use in a browser. This would've also enticed me to further check out Supabase functionality & features and I may still look at it in the future, overall I'm always happy to learn about new projects especially those that put in hard work to replace commercial tools with opensource ones.

Again, I understand trying to promote your own software and I also understand that this blog post was pretty much created as a log of 'how we did it for our own thing' but without any real reason for using Supabase for this specifically, whether that be a reason you chose to use it in your own implementation (other than moar dogfood) or a reason that a user reading the blog and attempting their own implementation, it feels a bit empty/unnecessary.

I don't really know how to explain it which is why I'm failing here at this; I don't want to come off as an ass as I did appreciate (and benefit from) the core information provided. I was also a bit excited to learn about this project in the first place. However, now I'm left at a point where I've done the complete implementation and more and not only has Supabase in this case provided zero value or extra features, it's just been more of a hindrance. IIRC, the article/video doesn't even mention that only a standard pg client library is needed for this and that you're just using Supabase just 'cuz; which again, I totally understand in your own project.

Again I don't want to come off as an ass or discount anyone's hard work, including the post I've benefited from, just trying to explain my experience as it cost me some time and frustration.

Some clear and up-front messaging on the lack of SQLability and the reasons for that decision (whether people agree or not is an aside at this point) and perhaps recommendation for solutions for those requiring that functionality would be great and it seems like this is now a goal, so glad to see that.

In reference to the particular YT video + blog post I mentioned, again it's tough to even provide positive/constructive criticism as it seems like whomever was implementing it simply logged/described what they were doing in their own project; they didn't have to do this, blog it or share it and the code - regardless of my confusion as to what the point of using the Supbase client libraries in this example are, I did benefit from the information and am thankful for it; but I still feel compelled to mention that it would've been great if at least there was a note at the top of the blog and/or in the video mentioning something like "No special Supabase features/functionality above/beyond what a typical pg client library is used in this implementation and the same result can be achieved w/out using Supabase; we're using it because we can, we _are_ Supabase after all" or something; TBH even if there was just a single major feature used/implemented in this example to show off Supbase functionalities or something then it'd be totally valid as a promotional article but as it stands it feels like some weird in-between (again, I totally get it's written like whomever was implementing it was mainly b/logging their progress/implementation but others may see it less this way and be even more put off on the eventual realization)

Again, please take this with **full kindness and appreciation**. There's some harsher-seeming replies in this thread (from over 2 years ago none-the-less and I as well dug for more time than I should've had to get the answer this wasn't possible and by seeing this thread no-less which contains some annoyed posts which for some would be even more off-putting) and I don't want to see your hard work stomped on due to people expecting (and IMO, not unreasonably in this case) something that's not a thing and then confusedingly and fiendishly searching docs for a non-existent API call eventually resorting to Google to eventually find the first hint at an answer, a few-year-old thread with other annoyed developers saying (admittedly not very helpful) things like "I quit here".

**Reply by steve-chavez** [reply]
> This is also when I realized; what's the point of even using Supabase here? I get showing off your own product and dogfooding it and considering that the blog was based off your internal implementation but for the purposes of the implementation in the tutorial, from what I can tell, it's using ZERO Supabase-specific features and adding ZERO value in this case, in fact it's almost the opposite, it's only more modules to be loaded in node, more services to be run, etc.

The [pgvector blog post](https://supabase.com/blog/openai-embeddings-postgres-vector) runs on [Edge Functions](https://supabase.com/docs/guides/functions). Using a direct PostgreSQL connection in this environment will quickly get you a `FATAL too many connections` error in PostgreSQL(since there is a connection limit), causing all subsequent function calls to fail.

For solving this you need a connection pool, and this has to run in another server since it has to be persistent. 

So the immediate benefits of using the `supabase-js` library for the blog post:

- Builtin connection pool. Doing this in a custom server would bring more complexity: libraries, tuning, etc. What's the correct pool size? Does the pool library enforce a lifetime for the connections? Does it timeout when acquiring a connection fails? There are many failure points really, and all of this is already solved when using `supabase-js`.  More details and more benefits [here](https://supabase.com/docs/guides/database/connecting-to-postgres#direct-vs-pooling-vs-api).
- Later on, you'll be able to use PostgreSQL row-level security plus many other features. See [here](https://supabase.com/docs/guides/api#rest-api-overview).

**Reply by ForbiddenEra** [reply]
Thanks for the clarification :)

**Reply by matart15** [reply]
After using a supabase with RPC and VIEWS, we realized there is 1 big problem

both RPC and VIEWS are separate from our codebase. this big problem introduces series of small problems. 
- We have to manualliy check all server enviroiments have same function logic
- If someone mistakingly deletes or edits function things go BOOM. With SQL inside codebase, function changes go through code review.
- extra steps on deployment. ( or extra script for CI/CD ) which is bad and source of problems

At this point we want to connect Prisma or Drizzle directly to our DB. Which makes supabase useless. 

If we use Drizzle, I am not sure what advantage `supabase + drizzle` have over `planetscale + drizzle`.

Let me know If there is a way to place my RPC and VIEWs inside codebase and deploy automatically without wrestling CI/CD

**Reply by ThingEngineer** [reply]
100% agree with your codebase issue. This is our biggest pain point and the only reason we use other methods to connect to our SB postgres like drizzle along side some of our SB projects; for drizzle.execute(sql`SELECT...`) vs SB.rpc('get_stuff') on a stored procedure. Maintaining two connections is not ideal but much less of an issues than having code outside of the repository and having to run a migration anytime a stored procedure is changed.

We get the "reason" for the lack of functionality (executing raw sql) is partly that no one should ever run raw sql in the client and much of the api is client first. That said, there is still a lot of the api that is meant for server side only. We had hopes this would change with the new ssr package but no. 

!!! UPDATE: I was wrong, I spoke too soon. Shortly after posting this I dove into Postgres functions and saw all that I had been missing. It has been and is the most powerful skill addition and I wish I had found it decades ago. Yes my code is outside, but with edge functions that is just the natural evolution. I obviously have all the migration files so it's all here just in a different place. After a full overhaul on a large app it now feels like the db is running locally. The encapsulation of complex logic, moving it to the db server, batch operations, read replication and Supabase's performance advisors have been a literal +10x speed boost. And the security, I'll just stop here and say again, I was wrong. I love Postgres functions and supabase.rpc - if you are trying to "Execute SQL in Supabase API" and are failing, that's a sign it is time to adapt, learn and adopt.

**Reply by GSeyfried** [reply]
I am trying to do something like this to execute commands, I could not figure out the issue until i realized it's just not possible:

`def execute_sql_command(supabase: Client, command: str,params=None):
    """
    Execute a SQL command using Supabase by calling a predefined PostgreSQL function.
    """
    sql_command = sql_commands.get(command)
    try:
        response = supabase.rpc('execute_sql_command', {'sql_command': sql_command}).execute()
        if response.error:
            raise Exception(response.error.message)
        st.error("SQL command '{command}' executed successfully.")
    except Exception as e:
        st.error(f"An error occurred while executing the SQL command: {str(e)}")`

**Reply by elyobo** [reply]
@matart15 if you're using migrations ([as recommended](https://supabase.com/docs/guides/platform/maturity-model#in-production)) then your RPCs and views would be in your code and subject to review, wouldn't they?

**Reply by matart15** [reply]
@elyobo To answer for you question `Yes migration solves problem I wrote before`.

At the time of writing We were not using migration very much. 

Now we used it more, migraion have own small prooblem

-  imperative migrations are hard to manage. it creates too many files with ALTER TABLE, REPLACE FUNCTION. ( gues it should be solved with squash. but squash was hard to do and dangerous if you don't fully understand what you are doing ).
Drizzle have same problem ( I guess ). prisma solves it with own declerative prisma language.

I am not complaining. Just wishing there were a simple solution.

**Reply by ox-harris** [reply]
Have had my fair share of a hair-pulling experience here. So I can relate. And no! Prisma, drizzle, or other ORM/migration tool can't remove the inherently tricky and complex nature of migrations. The industry has yet to find a better practice than migrations as we know it.

That said, this is also the problem I am working on with [Linked QL](https://github.com/linked-db/linked-ql). Its a simple query client that has a simplistic, declarative take on migrations. I think you'll love it.

 If you aren't too far along with other approaches, you might want to look that up. Hopefully that helps.

**Reply by matart15** [reply]
@ox-harris nice work. Auto-Versioning looks interesting. I will try it.

**Reply by ox-harris** [reply]
Great. Happy to help with anything when you do try it out.

**Reply by steve-chavez** [reply]
> After using a supabase with RPC and VIEWS, we realized there is 1 big problem
> both RPC and VIEWS are separate from our codebase. this big problem introduces series of small problems.

You can put your SQL files in source control, overall migrations nowadays are better handled with "declarative schemas". See:

- https://supabase.com/docs/guides/local-development/declarative-database-schemas
- https://supabase.com/blog/declarative-schemas

**Reply by andrastoth-ws** [reply]
It baffles me when any company decides to shrink the rich language of SQL into a phrase book basically. When you can pass in raw SQL statements from, say, JavaScript, you _augment_ the capabilities of both languages.

If you do joins and calculations on the backend because somebody hidden a basic SQL feature from you that could have been done in 1 query, you already wasted a ton of performance, for what exactly? From a formal language point of view, having the ability to execute raw queries would mean you lift JS into a multi-language system where JS (or other language) controls the control flow... Raw queries allow the full declarative grammar of SQL to be available in an imperative host!

Just for a connection pool I don't want to throw away the expressivity of a mature and feature-full language. Terrible design decision!

(There are players like `slonik` who embrace and enhance the raw SQL and they also provide their own connection pool. I worked with it and it was a great experience - I could run queries from inside my IDE on the dev cloud and so on, without even starting the backend.)

**Reply by steve-chavez** [reply]
> There are players like slonik who embrace and enhance the raw SQL and they also provide their own connection pool.

The supabase-js library can be used from the frontend unlike any other ORM, this is why it has this constraint. Maybe worth repeating the first comment in this thread:

> SQL cannot be fully exposed to frontend clients because it opens a door for forcing inefficient/slow queries that can starve your database of resources. Also a malicious user could do a Denial-of-Service.
> https://github.com/orgs/supabase/discussions/3419#discussioncomment-1428169

**Comment by Johnrobmiller** [comment]
Thank you for posting this. Is there anybody who is able to clarify what it means to "generate a view"?

**Reply by Johnrobmiller** [reply]
Answering my own question: https://supabase.com/blog/postgresql-views

**Reply by MessiDaGod** [reply]
Hey have you heard of ChatGPT?

**Reply by Johnrobmiller** [reply]
> Hey have you heard of ChatGPT?

I don't think chatGPT was launched when I posted the original comment, but I don't think your responses deserves 222 downvotes -- that seems excessive. I'm not so fragile that I can't take a little heat every now and then 😅

**Reply by elyobo** [reply]
It was an unhelpful and rude comment. The thumbs down seem like a mild and reasonable response.

**Reply by qubydev** [reply]
> Hey have you heard of ChatGPT?

Grow up, kid.
If you believe ChatGPT can solve all your problems, then either you’re unaware of complex projects or you’re lying.

**Comment by revskill10** [comment]
Any udpate on this to use `supabase-js` to run raw sql query ?

**Reply by abdirahmn1** [reply]
here's a medium article showing how you can execute a custom sql statement by defining your logic in a postgres function before calling it with `supabase.rpc('function-name')`

https://medium.com/@razvanst/how-to-run-custom-sql-queries-using-functions-in-supabase-f81bfab780a7

**Reply by revskill10** [reply]
@abdirahmn1 I know about rpc. But please upvote to have built-in capability with supabase-js instead of managing tons of rpc functions.

**Reply by abdirahmn1** [reply]
You can use `node-postgres` to connect to your supabase DB and execute your SQL queries that way. IDK why you are hell bent on `supabase-js` having to support a way to execute raw SQL 🤷‍♂️

here's an article from supabase - 

https://supabase.com/docs/guides/database/connecting-to-postgres

**Reply by revskill10** [reply]
@abdirahmn1 I deploy on Vercel on the edge.

Tell me how to run raw sql there with your node-postgres ?

**Reply by abdirahmn1** [reply]
I HATE to be that guy but please rtfm 👀, sorry if I'm being rude.

https://supabase.com/docs/guides/database/connecting-to-postgres

the docs have implemented a solution to your problem using an `ORM` and also the `postgres` lib.

`node-postgres` achieves practically the same thing that `postgres` does. here are the docs to `node-postgres` -

https://node-postgres.com/

stay awesome 🙂🤝

**Comment by davidosunkeye** [comment]
Can Supabase team just create wrapperJs to write raw SQL. I am moved from FireBase and I am just lost why it’s hard to achieve raw SQL on Supabase

**Reply by abdirahmn1** [reply]
here's a medium article showing how you can execute a custom sql statement by defining your logic in a postgres function before calling it with `supabase.rpc('function-name')`

https://medium.com/@razvanst/how-to-run-custom-sql-queries-using-functions-in-supabase-f81bfab780a7

**Reply by andredifs** [reply]
I just want to executre raw SQL, is it really so hard to make a built-in method for this? 🥲

**Reply by matart15** [reply]
> I just want to executre raw SQL, is it really so hard to make a built-in method for this? 🥲

@andredifs My understanding is `just executre raw SQL` is hard because of `security`. Supabase client is designed to run in client side.

**Comment by louis030195** [comment]
Anyone figured out how to execute DYNAMIC queries?

I'm generating SQL queries using a LLM so I can't hard code the query in a function, perfect solution would be to have a function that let the user execute read only queries (don't care about RLS stuff)

```sql
CREATE OR REPLACE FUNCTION execute_raw_select_sql(query text)
RETURNS SETOF RECORD AS $$
DECLARE
    forbidden_pattern text[] := ARRAY['INSERT', 'UPDATE', 'DELETE', 'TRUNCATE', 'ALTER', 'DROP', 'CREATE', 'REVOKE', 'GRANT'];
BEGIN
    -- IF EXISTS (SELECT 1 FROM unnest(forbidden_pattern) pattern WHERE upper(query) LIKE '%' || upper(pattern) || '%') THEN
    --     RAISE EXCEPTION 'Query contains forbidden operations';
    -- END IF;
    -- TODO: the if is broken somehow

    RETURN QUERY EXECUTE query;
END;
$$ LANGUAGE plpgsql;
```

Getting this

```
{
  code: '0A000',
  details: null,
  hint: null,
  message: 'materialize mode required, but it is not allowed in this context'
}
```

Any help appreciated 🙏

My workaround is to generate and execute JS code instead :/ pref SQL

**Reply by bahtiaryusuf10** [reply]
is this issue resolved now bro?

**Comment by ctwhome** [comment]
you can create a function that allows you to pass your raw SQL query as a parameter. Just be aware of the security implications it might bring. 

```sql
CREATE OR REPLACE FUNCTION execute_dynamic_sql(sql_command TEXT)
RETURNS void AS $$
BEGIN
  EXECUTE sql_command;
END;
$$ LANGUAGE plpgsql;
```

and to run:
```
SELECT execute_dynamic_sql('UPDATE your_table_name SET your_data]');
```

mote that the function returns void, meaning nothing. if you want the function to return some data you need to change the function return from void to something you need like a JSON response

**Reply by hieuvo** [reply]
I tried the suggestion above, but any query with join or `with as` still produces error `materialize mode required, but it is not allowed in this context`
```sql
CREATE OR REPLACE FUNCTION execute_sql(query TEXT)
RETURNS SETOF RECORD AS $$
BEGIN
  RETURN QUERY EXECUTE query;
END;
$$ LANGUAGE plpgsql;
```
```sql
SELECT MAX(date) AS max_date
FROM checkins c1
WHERE c1.user_id = '6679b002-f0a8-47b1-84ba-d73a56d47fe5'
  AND NOT EXISTS (
      SELECT 1
      FROM checkins c2
      WHERE c2.user_id = c1.user_id
        AND c2.date = c1.date + INTERVAL '1 day'
  );
```

**Reply by just-abdou** [reply]
I was able to find a walk around it if ur still interested(sorry for the late response)
```
set check_function_bodies = off;

CREATE OR REPLACE FUNCTION public.sql_query_runner(query text)
 RETURNS text
 LANGUAGE plpgsql
AS $function$
DECLARE
    result TEXT;
BEGIN
    BEGIN
        EXECUTE 'SELECT array_to_string(array_agg(row_to_json(t)::text), '','') FROM (' || query || ') t'
        INTO result;

        RAISE NOTICE 'Query result: %', result;

        RETURN result;

    EXCEPTION
        WHEN OTHERS THEN
            RAISE NOTICE 'An error occurred: %', SQLERRM;
            RAISE NOTICE 'Failed query: %', query;

            -- You can choose to return NULL or re-raise the exception
            RETURN SQLERRM;
    END;
END;
$function$
;

```
 you can then decode the string in the client side
